### PR TITLE
Can storage

### PR DESCRIPTION
**Pull request by SuksAE**
Can storage generator to satisfy issue #330 (https://github.com/florianfesti/boxes/issues/330).

Note: depends on fillHole function, which is a different pull request...

### CODE REVIEW

File: boxes/generators/can_storage.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        if self.top_edge != "e":
+            self.corner(90, self.thickness)
+            self.edge(0.5 * self.canDiameter)            
+            self.corner(-90, 0.25 * self.canDiameter)
+        else:
+            self.moveTo(-self.burn, self.canDiameter + self.thickness, -90)
+            self.corner(90, 0.25 * self.canDiameter)
+            self.edge(self.thickness)
+
+        self.edge(0.5 * self.canDiameter - self.thickness)
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:34]: May be have a look at the `.polyline()` method.

File: boxes/generators/can_storage.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            type=ArgparseEdgeType("eEš"), choices=list("eEš"),
+            default="š", help="edge type for bottom edge")
+
+        # Add non default cli params if needed (see argparse std lib)
+        self.argparser.add_argument(
+            "--canDiameter",  action="store", type=float, default=67,
+            help="outer diameter of the cans to be stored (in mm)")
+        self.argparser.add_argument(
+            "--canHight",  action="store", type=float, default=115,
+            help="hight of the cans to be stored (in mm)")
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:77]: I wonder if there should be a parameter for the number of rows. But having just two as a default is fine for now.

File: boxes/generators/can_storage.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            # fill with holes
+            border = [
+                (0, 0), 
+                (self.top_chute_depth, 0), 
+                (self.top_chute_depth, self.width), 
+                (0, self.width),                
+                ]
+            if self.mydebug:
+                self.showBorderPoly(border)
+            if self.fillHoles_fill_pattern != "no fill":
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:93]: This is a sign of a bad API. `.fillHoles` should just handle the "no fill" option by doing nothing.

File: boxes/generators/can_storage.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                    border=border, 
+                    max_radius=self.fillHoles_hole_max_radius, 
+                    hspace=self.fillHoles_space_between_holes, 
+                    bspace=self.fillHoles_space_to_border, 
+                    min_radius=self.fillHoles_hole_min_radius, 
+                    style=self.fillHoles_hole_style,
+                    maxcnt=2000)            
+        
+    def render(self):
+        self.mydebug = False # set to True to enable debugging
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:190]: Please just use the regular debug parameter. It was added for exactly this purpose.

File: boxes/generators/can_storage.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    def cb_top_chute(self, nr):
+        if nr == 0:
+            # fill with holes
+            border = [
+                (0, 0), 
+                (self.top_chute_depth, 0), 
+                (self.top_chute_depth, self.width), 
+                (0, self.width),                
+                ]
+            if self.mydebug:
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:91]: Why isn't fillHoles just doing the debug output?

File: boxes/generators/can_storage.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                self.showBorderPoly(border)
+            if self.fillHoles_fill_pattern != "no fill":
+                self.fillHoles(
+                    pattern="hbar",
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:95]: The long slots make those chutes very fragile especially of the cans don't rest on the edges of the chute that are connected to the side walls. Beverages are heavy and you may run into all kind of trouble with the chute bending or even breaking.

File: boxes/generators/can_storage.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+                    hspace=2, 
+                    bspace=2, 
+                    )            
+            
+    def cb_sides(self, nr):
+        if nr == 0:
+            # for debugging only
+            if self.mydebug:
+                # draw orientation points
+                self.hole(0, 0, 1, color=Color.ANNOTATIONS)
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:127]: This looks like it should be a loop

File: boxes/generators/can_storage.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            # bottom chute
+            with self.saved_context():
+                self.moveTo(0, 0.5 * self.thickness, self.angle)
+                self.fingerHolesAt(0, 0, self.depth / math.cos(math.radians(self.angle)), 0)
+            # top chute
+            with self.saved_context():
+                self.moveTo(0, self.thickness + self.canDiameter + self.bottom_chute_height + self.top_chute_height + 0.5 * self.thickness, -self.angle)
+                self.fingerHolesAt(0, 0, self.top_chute_depth, 0)
+            # front barrier
+            with self.saved_context():
```
Comment (florianfesti) [inline boxes/generators/can_storage.py:160]: The front barrier needs to be at least one thickness in to provide the necessary strength. Having an f edge connection here is a bad idea. All the cans are going to roll and hit this with momentum.

**Review by florianfesti — COMMENTED** [review]
This is a really nice generator and one can clearly see how much work went into this.

But before this can be merged one needs to actually cut it and try with real cans. The cans going around the corner is a tricky thing and small mistakes can easily get them stuck. It is quite possible that more angle surfaces are needed to make that work. Or it is just fine. How knows?

File: boxes/generators/can_storage.py (hunk 3 above)
Comment (SuksAE) [reply boxes/generators/can_storage.py:93]: ---> fixed in PR #400

File: boxes/generators/can_storage.py (hunk 4 above)
Comment (SuksAE) [reply boxes/generators/can_storage.py:190]: --> done

File: boxes/generators/can_storage.py (hunk 6 above)
Comment (SuksAE) [reply boxes/generators/can_storage.py:95]: changed in PR #400

File: boxes/generators/can_storage.py (hunk 8 above)
Comment (SuksAE) [reply boxes/generators/can_storage.py:160]: --> done

**Comment by florianfesti** [comment]
The stackable edges kinda suggest the dispensers can be stacked. But there is nothing that keeps them in place left to right. Adding stackable edges to the back plate will help with that. On the front one will need small pieces to glue on the inside of the feet to prevent them from slipping.

**Comment by HaSHsss** [comment]
image
This part really upsets me.
It makes it difficult to take cans easily.
You have to pretend to be a crab with claws to hook cans.

**Comment by SuksAE** [comment]
Maybe you want to give it a try?
I hope this doesn`t weaken the design too much...

**Comment by florianfesti** [comment]
May be testing it in a smaller size is quicker and cheaper. May be in a similar size to this:  https://www.thingiverse.com/thing:898062

**Comment by SuksAE** [comment]
Spoiler alert:

CanStorage_AA_cut


CanStorage_AA

Designing a battery holder with CanStorage is not so easy - AA size is a lot harder to manage than 400g cans. But I kind of like the result...

**Comment by SuksAE** [comment]
from my point of view this is ready to be merged into master...

**Comment by florianfesti** [comment]
Sorry for the slow response. I was on vacation, then sick, then sick on vacation. I hope we can wrap this up soon.

Minor thing: The last patch misses changes to `static/samples/samples.sha256`. See https://florianfesti.github.io/boxes/html/CONTRIBUTING.html#provide-photos-for-generators for details. If you don't have the right tooling at hand I can also generate the thumbnails for you and amend the patch.

I am wondering whether we should squash the patches or whether it is worth keeping the different development steps in separate patches. Any opinion on that?

**Comment by SuksAE** [comment]
OK, I missed that with the pictures - if you could handle this, it would be fine.

Regarding the patches - I don't see any benefit in keeping the separate steps. I think the end result is most important outcome, and as there was nothing like a release to the public in between, the intermediate steps can be omitted...

**Comment by florianfesti** [comment]
Merged as 43bd7d12d318e353cc1178f714701a23ef6ed989

Thanks for all the work and patience!

**Comment by SuksAE** [comment]
Final question: when will this be available on www.festi.info/boxes.py?
Is there an update schedule for the web server or how do you handle the updates?

**Comment by florianfesti** [comment]
I am updating the web instance by hand. I am always a bit hesitant to do that and will typically wait until I do have enough time at hand to deal with any issues that may come up. Unfortunately the docker service on this rather old OS isn't too reliable and sometimes needs a love bump.

Anyway, updated.

**Comment by AshaiRey** [comment]
Too ALL worked on this can generator.
Thank you for your hard work to make this possible.
I really appreciate it very much.

**Pull request closed without merging**